internal/experiment/openvpn/openvpn_test.go

```diff
@@ -0,0 +1,421 @@
+		}
+	})
+	t.Run("providers with userpass auth method raise error, not yet implemented", func(t *testing.T) {
+		m := openvpn.NewExperimentMeasurer(openvpn.Config{}, "openvpn").(openvpn.Measurer)
+		ctx := context.Background()
+		sess := makeMockSession()
+		opts, err := m.GetCredentialsFromOptionsOrAPI(ctx, sess, "tunnelbear")
+		if !errors.Is(err, openvpn.ErrInvalidInput) {
+			t.Fatalf("expected err=ErrInvalidInput, got %v", err)
+		}
+		if opts != nil {
+			t.Fatal("expected opts=nil")
+		}
+	})
+	t.Run("known cert auth provider and creds in options is ok", func(t *testing.T) {
+		config := openvpn.Config{
+			SafeCA:   "base64:Zm9v",
+			SafeCert: "base64:Zm9v",
+			SafeKey:  "base64:Zm9v",
+		}
+		m := openvpn.NewExperimentMeasurer(config, "openvpn").(openvpn.Measurer)
+		ctx := context.Background()
+		sess := makeMockSession()
+		opts, err := m.GetCredentialsFromOptionsOrAPI(ctx, sess, "riseup")
+		if err != nil {
+			t.Fatalf("expected err=nil, got %v", err)
+		}
+		if opts == nil {
+			t.Fatal("expected non-nil options")
+		}
+	})
+	t.Run("known cert auth provider and bad creds in options returns error", func(t *testing.T) {
+		config := openvpn.Config{
+			SafeCA:   "base64:Zm9v",
+			SafeCert: "base64:Zm9v",
+			SafeKey:  "base64:Zm9vaaa",
+		}
+		m := openvpn.NewExperimentMeasurer(config, "openvpn").(openvpn.Measurer)
+		ctx := context.Background()
+		sess := makeMockSession()
+		opts, err := m.GetCredentialsFromOptionsOrAPI(ctx, sess, "riseup")
+		if !errors.Is(err, openvpn.ErrBadBase64Blob) {
+			t.Fatalf("expected err=ErrBadBase64, got %v", err)
+		}
+		if opts != nil {
+			t.Fatal("expected nil opts")
+		}
+	})
+	t.Run("known cert auth provider with null options hits the api", func(t *testing.T) {
+		config := openvpn.Config{}
+		m := openvpn.NewExperimentMeasurer(config, "openvpn").(openvpn.Measurer)
+		ctx := context.Background()
+		sess := makeMockSession()
+		opts, err := m.GetCredentialsFromOptionsOrAPI(ctx, sess, "riseup")
+		if err != nil {
+			t.Fatalf("expected err=nil, got %v", err)
+		}
+		if opts == nil {
+			t.Fatalf("expected not-nil options, got %v", opts)
+		}
+	})
+	t.Run("known cert auth provider with null options hits the api and raises error if api fails", func(t *testing.T) {
+		config := openvpn.Config{}
+		m := openvpn.NewExperimentMeasurer(config, "openvpn").(openvpn.Measurer)
+		ctx := context.Background()
+
+		someError := errors.New("some error")
+		sess := makeMockSession()
+		sess.MockFetchOpenVPNConfig = func(context.Context, string, string) (*model.OOAPIVPNProviderConfig, error) {
+			return nil, someError
+		}
+
+		opts, err := m.GetCredentialsFromOptionsOrAPI(ctx, sess, "riseup")
+		if !errors.Is(err, someError) {
+			t.Fatalf("expected err=someError, got %v", err)
+		}
+		if opts != nil {
+			t.Fatalf("expected nil options, got %v", opts)
+		}
+	})
+}
+
+func TestAddConnectionTestKeys(t *testing.T) {
+	t.Run("append connection result to empty keys", func(t *testing.T) {
```

Comment:
Please, add a test for the UDP case as well

Reply:
added